source/slang/slang-check-decl.cpp

```diff
@@ -2101,6 +2101,19 @@ namespace Slang
             {
                 getSink()->diagnose(varDecl->type.exp->loc, Diagnostics::incompleteTypeCannotBeUsedInBuffer, elementType);
             }
+            if (doesTypeHaveTag(elementType, TypeTag::Unsized))
+            {
+                // If the element type is unsized, it can only be an array of resource types that we can legalize out.
+                // Ordinary unsized arrays are not allowed in a constant buffer since we cannot translate it to
+                // valid HLSL or SPIRV.
+                ArrayExpressionType* trailingArrayType = nullptr;
+                VarDeclBase* trailingArrayField = getTrailingUnsizedArrayElement(elementType, varDecl, trailingArrayType);
```

Comment:
Would it be better to assign a temporary modifier to parents of ::Unsized elements for when we add an initial Unsized element so we can skip this check?

Reply:
We could, but this happens infrequent enough so I didn't bother adding that complexity in.